blueprints/multi-interface-servers.md

```diff
@@ -0,0 +1,670 @@
+listens for connections.
+
+The specific use-case outlined above, obviously, entails some management be done
+on the cache servers to actually support multiple interfaces answering to the
+same "main" IP address and this should be considered outside of
+<abbr title="Apache Traffic Control">ATC</abbr>'s scope of operation.
+
+## Data Model Impact
+The proposed structure of each of these interfaces is given below as a pair of
+TypeScript<sup>[1](#typescript)</sup> interfaces:
+```typescript
+interface IPAddress {
+	/**
+	 * The actual IP (v4 or v6) address which is used by an interface.
+	 * Any "mask" should be present here as a CIDR-notation suffix.
+	 */
+	address: string;
+
+	/**
+	 * The IP (v4 or v6) address of the gateway used by this IP address.
+	 */
+	gateway: string;
+
+	/**
+	 * Tells whether or not this address of this interface is the server's
+	 * "service" address.
+	 * At least one address of EXACTLY ONE interface MUST have this set to
+	 * 'true' for a server.
+	 */
+	serviceAddress: boolean;
+}
+
+interface Interface {
+	/**
+	 * The name of the interface device on the server e.g. eth0.
+	 */
+	name: string;
+	/**
+	 * These will be all of the IPv4/IPv6 addresses assigned to the interface,
+	 * including gateways and "masks".
+	 * It is illegal for an interface to not have at least one associated IP
+	 * address.
+	 */
+	ipAddresses: Array<IPAddress> & {0: IPAddress};
+	/**
+	 * Whether or not Traffic Monitor should monitor this particular interface.
+	 */
+	monitor: boolean;
+	/**
+	 * The interface's Maximum Transmission Unit.
+	 * If this is 'null' it is assumed that the interface's MTU is not known/is
+	 * irrelevant.
+	 */
+	mtu: 1500 | 9000 | null;
+}
+```
+We don't have a real data model for ATC, so what follows is an approximately
+defined representation of a "server" that is valid within certain contexts. The
+modifications made to it for the purposes of the changes herein proposed should
+be considered valid in those same contexts - whatever they may be.
+
+```typescript
+interface Server {
+
+	cachegroup?:  string | null;
+	cachegroupId: number;
+	cdnId:        number;
+	cdnName?:     string | null;
+	domainName:   string;
+	guid?:        string | null;
+	hostName:     string;
+	httpsPort?:   number | null;
+	id?:          number;
+
+	// ILO things aren't being moved in because they also require a
+	// username/password that aren't used by any other interface type.
+	iloIpAddress?: string | null;
+	iloIpGateway?: string | null;
+	iloIpNetmask?: string | null;
+	iloPassword?:  string | null;
+	iloUsername?:  string | null;
+
+	/**
+	 * New field containing additional interfaces.
+	 * It is illegal for a server to not have at least one interface.
+	 */
+	interfaces: Array<Interface> & {0: Interface};
+	lastUpdated?:  string;
+
+	// notice not here: mgmtIpAddress, mgmtIpGateway, mgmtIpNetmask
+	// these will be moved into the interfaces property
```

Comment:
Where?  Should there be an purpose field on an interface?

Reply:
This would mean you could also model out-of-band management the same way as mgmtIpAddress etc.

Reply:
It would need to be associated with the username and password information currently on the server object, which is why the decision was made to leave it be, as no other interface would ever use those fields.

Reply:
As far as "purpose" fields... ¯\\\_(ツ)\_/¯

could be useful, but I worry that they'd fill up with mostly junk like "interface eth0". Maybe if it was optional, and definitely with a character limit. Like 60 characters, or something.

Reply:
How do you represent a server with two public and two private interface (used for different private networks)?

Reply:
There's no distinction between "public" and "private" interfaces, if that's what your asking

Reply:
You distinguish those using the "monitored" field.  If an interface is monitored it is a "public" interface, if it is not monitored it is a "private" interface.